Subject: [PATCH] pure motion
---
Index: facefusion/wording.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/facefusion/wording.py b/facefusion/wording.py
--- a/facefusion/wording.py	(revision d1385fed0d45329171c2119de5d3713a09d18399)
+++ b/facefusion/wording.py	(date 1750151355203)
@@ -184,6 +184,7 @@
 		'frame_enhancer_model': 'choose the model responsible for enhancing the frame',
 		'frame_enhancer_blend': 'blend the enhanced into the previous frame',
 		'lip_syncer_model': 'choose the model responsible for syncing the lips',
+		'lip_syncer_pure_motion': 'refine with live-portrait',
 		'lip_syncer_weight': 'specify the degree of weight applied to the lips',
 		# uis
 		'open_browser': 'open the browser once the program is ready',
@@ -310,6 +311,7 @@
 		'job_runner_job_action_dropdown': 'JOB ACTION',
 		'job_runner_job_id_dropdown': 'JOB ID',
 		'lip_syncer_model_dropdown': 'LIP SYNCER MODEL',
+		'lip_syncer_pure_motion_slider': 'PURE MOTION',
 		'lip_syncer_weight_slider': 'LIP SYNCER WEIGHT',
 		'log_level_dropdown': 'LOG LEVEL',
 		'output_audio_encoder_dropdown': 'OUTPUT AUDIO ENCODER',
Index: facefusion/uis/components/lip_syncer_options.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/facefusion/uis/components/lip_syncer_options.py b/facefusion/uis/components/lip_syncer_options.py
--- a/facefusion/uis/components/lip_syncer_options.py	(revision d1385fed0d45329171c2119de5d3713a09d18399)
+++ b/facefusion/uis/components/lip_syncer_options.py	(date 1750151355193)
@@ -10,11 +10,13 @@
 from facefusion.uis.core import get_ui_component, register_ui_component
 
 LIP_SYNCER_MODEL_DROPDOWN : Optional[gradio.Dropdown] = None
+LIP_SYNCER_PURE_MOTION_SLIDER : Optional[gradio.Slider] = None
 LIP_SYNCER_WEIGHT_SLIDER : Optional[gradio.Slider] = None
 
 
 def render() -> None:
 	global LIP_SYNCER_MODEL_DROPDOWN
+	global LIP_SYNCER_PURE_MOTION_SLIDER
 	global LIP_SYNCER_WEIGHT_SLIDER
 
 	has_lip_syncer = 'lip_syncer' in state_manager.get_item('processors')
@@ -23,6 +25,14 @@
 		choices = processors_choices.lip_syncer_models,
 		value = state_manager.get_item('lip_syncer_model'),
 		visible = has_lip_syncer
+	)
+	LIP_SYNCER_PURE_MOTION_SLIDER = gradio.Slider(
+		label = wording.get('uis.lip_syncer_pure_motion_slider'),
+		value = state_manager.get_item('lip_syncer_pure_motion'),
+		step = calc_float_step(processors_choices.lip_syncer_pure_motion_range),
+		minimum = processors_choices.lip_syncer_pure_motion_range[0],
+		maximum = processors_choices.lip_syncer_pure_motion_range[-1],
+		visible = has_lip_syncer
 	)
 	LIP_SYNCER_WEIGHT_SLIDER = gradio.Slider(
 		label = wording.get('uis.lip_syncer_weight_slider'),
@@ -33,21 +43,23 @@
 		visible = has_lip_syncer
 	)
 	register_ui_component('lip_syncer_model_dropdown', LIP_SYNCER_MODEL_DROPDOWN)
+	register_ui_component('lip_syncer_pure_motion_slider', LIP_SYNCER_PURE_MOTION_SLIDER)
 	register_ui_component('lip_syncer_weight_slider', LIP_SYNCER_WEIGHT_SLIDER)
 
 
 def listen() -> None:
 	LIP_SYNCER_MODEL_DROPDOWN.change(update_lip_syncer_model, inputs = LIP_SYNCER_MODEL_DROPDOWN, outputs = LIP_SYNCER_MODEL_DROPDOWN)
+	LIP_SYNCER_PURE_MOTION_SLIDER.change(update_lip_syncer_pure_motion, inputs = LIP_SYNCER_PURE_MOTION_SLIDER)
 	LIP_SYNCER_WEIGHT_SLIDER.release(update_lip_syncer_weight, inputs = LIP_SYNCER_WEIGHT_SLIDER)
 
 	processors_checkbox_group = get_ui_component('processors_checkbox_group')
 	if processors_checkbox_group:
-		processors_checkbox_group.change(remote_update, inputs = processors_checkbox_group, outputs = [ LIP_SYNCER_MODEL_DROPDOWN, LIP_SYNCER_WEIGHT_SLIDER ])
+		processors_checkbox_group.change(remote_update, inputs = processors_checkbox_group, outputs = [ LIP_SYNCER_MODEL_DROPDOWN, LIP_SYNCER_PURE_MOTION_SLIDER, LIP_SYNCER_WEIGHT_SLIDER ])
 
 
-def remote_update(processors : List[str]) -> Tuple[gradio.Dropdown, gradio.Slider]:
+def remote_update(processors : List[str]) -> Tuple[gradio.Dropdown, gradio.Slider, gradio.Slider]:
 	has_lip_syncer = 'lip_syncer' in processors
-	return gradio.Dropdown(visible = has_lip_syncer), gradio.Slider(visible = has_lip_syncer)
+	return gradio.Dropdown(visible = has_lip_syncer), gradio.Slider(visible = has_lip_syncer), gradio.Slider(visible = has_lip_syncer)
 
 
 def update_lip_syncer_model(lip_syncer_model : LipSyncerModel) -> gradio.Dropdown:
@@ -60,5 +72,11 @@
 	return gradio.Dropdown()
 
 
+def update_lip_syncer_pure_motion(lip_syncer_pure_motion : float) -> None:
+	lip_syncer_module = load_processor_module('lip_syncer')
+	lip_syncer_module.clear_inference_pool()
+	state_manager.set_item('lip_syncer_pure_motion', lip_syncer_pure_motion)
+
+
 def update_lip_syncer_weight(lip_syncer_weight : float) -> None:
 	state_manager.set_item('lip_syncer_weight', lip_syncer_weight)
Index: facefusion/uis/components/preview.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/facefusion/uis/components/preview.py b/facefusion/uis/components/preview.py
--- a/facefusion/uis/components/preview.py	(revision d1385fed0d45329171c2119de5d3713a09d18399)
+++ b/facefusion/uis/components/preview.py	(date 1750151355196)
@@ -132,6 +132,7 @@
 		'face_enhancer_weight_slider',
 		'frame_colorizer_blend_slider',
 		'frame_enhancer_blend_slider',
+		'lip_syncer_pure_motion_slider',
 		'lip_syncer_weight_slider',
 		'reference_face_distance_slider',
 		'face_selector_age_range_slider',
Index: facefusion/processors/modules/lip_syncer.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/facefusion/processors/modules/lip_syncer.py b/facefusion/processors/modules/lip_syncer.py
--- a/facefusion/processors/modules/lip_syncer.py	(revision d1385fed0d45329171c2119de5d3713a09d18399)
+++ b/facefusion/processors/modules/lip_syncer.py	(date 1750153796942)
@@ -1,6 +1,6 @@
 from argparse import ArgumentParser
 from functools import lru_cache
-from typing import List
+from typing import List, Tuple
 
 import cv2
 import numpy
@@ -20,10 +20,12 @@
 from facefusion.face_store import get_reference_faces
 from facefusion.filesystem import filter_audio_paths, has_audio, in_directory, is_image, is_video, resolve_relative_path, same_file_extension
 from facefusion.processors import choices as processors_choices
+from facefusion.processors.live_portrait import create_rotation, limit_expression
 from facefusion.processors.types import LipSyncerInputs, LipSyncerWeight
+from facefusion.processors.types import LivePortraitExpression, LivePortraitFeatureVolume, LivePortraitMotionPoints, LivePortraitPitch, LivePortraitRoll, LivePortraitScale, LivePortraitTranslation, LivePortraitYaw
 from facefusion.program_helper import find_argument_group
-from facefusion.thread_helper import conditional_thread_semaphore
-from facefusion.types import ApplyStateItem, Args, AudioFrame, BoundingBox, DownloadScope, Face, InferencePool, ModelOptions, ModelSet, ProcessMode, QueuePayload, UpdateProgress, VisionFrame
+from facefusion.thread_helper import conditional_thread_semaphore, thread_semaphore
+from facefusion.types import ApplyStateItem, Args, AudioFrame, BoundingBox, DownloadScope, DownloadSet, Face, InferencePool, ModelOptions, ModelSet, ProcessMode, QueuePayload, UpdateProgress, VisionFrame
 from facefusion.vision import read_image, read_static_image, restrict_video_fps, write_image
 
 
@@ -93,45 +95,173 @@
 			},
 			'type': 'wav2lip',
 			'size': (96, 96)
+		},
+		'live_portrait':
+		{
+			'hashes':
+			{
+				'feature_extractor':
+				{
+					'url': resolve_download_url('models-3.0.0', 'live_portrait_feature_extractor.hash'),
+					'path': resolve_relative_path('../.assets/models/live_portrait_feature_extractor.hash')
+				},
+				'motion_extractor':
+				{
+					'url': resolve_download_url('models-3.0.0', 'live_portrait_motion_extractor.hash'),
+					'path': resolve_relative_path('../.assets/models/live_portrait_motion_extractor.hash')
+				},
+				'eye_retargeter':
+				{
+					'url': resolve_download_url('models-3.0.0', 'live_portrait_eye_retargeter.hash'),
+					'path': resolve_relative_path('../.assets/models/live_portrait_eye_retargeter.hash')
+				},
+				'lip_retargeter':
+				{
+					'url': resolve_download_url('models-3.0.0', 'live_portrait_lip_retargeter.hash'),
+					'path': resolve_relative_path('../.assets/models/live_portrait_lip_retargeter.hash')
+				},
+				'stitcher':
+				{
+					'url': resolve_download_url('models-3.0.0', 'live_portrait_stitcher.hash'),
+					'path': resolve_relative_path('../.assets/models/live_portrait_stitcher.hash')
+				},
+				'generator':
+				{
+					'url': resolve_download_url('models-3.0.0', 'live_portrait_generator.hash'),
+					'path': resolve_relative_path('../.assets/models/live_portrait_generator.hash')
+				}
+			},
+			'sources':
+			{
+				'feature_extractor':
+				{
+					'url': resolve_download_url('models-3.0.0', 'live_portrait_feature_extractor.onnx'),
+					'path': resolve_relative_path('../.assets/models/live_portrait_feature_extractor.onnx')
+				},
+				'motion_extractor':
+				{
+					'url': resolve_download_url('models-3.0.0', 'live_portrait_motion_extractor.onnx'),
+					'path': resolve_relative_path('../.assets/models/live_portrait_motion_extractor.onnx')
+				},
+				'eye_retargeter':
+				{
+					'url': resolve_download_url('models-3.0.0', 'live_portrait_eye_retargeter.onnx'),
+					'path': resolve_relative_path('../.assets/models/live_portrait_eye_retargeter.onnx')
+				},
+				'lip_retargeter':
+				{
+					'url': resolve_download_url('models-3.0.0', 'live_portrait_lip_retargeter.onnx'),
+					'path': resolve_relative_path('../.assets/models/live_portrait_lip_retargeter.onnx')
+				},
+				'stitcher':
+				{
+					'url': resolve_download_url('models-3.0.0', 'live_portrait_stitcher.onnx'),
+					'path': resolve_relative_path('../.assets/models/live_portrait_stitcher.onnx')
+				},
+				'generator':
+				{
+					'url': resolve_download_url('models-3.0.0', 'live_portrait_generator.onnx'),
+					'path': resolve_relative_path('../.assets/models/live_portrait_generator.onnx')
+				}
+			}
+		},
+		'face_template':
+		{
+			'hashes':
+			{
+				'face_template':
+				{
+					'url': 'https://huggingface.co/bluefoxcreation/Templates/resolve/main/face-template/face_template.hash',
+					'path': resolve_relative_path('../.assets/templates/face_template.hash')
+				}
+			},
+			'sources':
+			{
+				'face_template':
+				{
+					'url': 'https://huggingface.co/bluefoxcreation/Templates/resolve/main/face-template/face_template.npy',
+					'path': resolve_relative_path('../.assets/templates/face_template.npy')
+				}
+			}
 		}
 	}
 
 
 def get_inference_pool() -> InferencePool:
 	model_names = [ state_manager.get_item('lip_syncer_model') ]
-	model_source_set = get_model_options().get('sources')
+	_, model_source_set = collect_model_downloads()
 
 	return inference_manager.get_inference_pool(__name__, model_names, model_source_set)
 
 
 def clear_inference_pool() -> None:
-	model_names = [ state_manager.get_item('lip_syncer_model') ]
+	model_names = [ state_manager.get_item('lip_syncer_model')]
 	inference_manager.clear_inference_pool(__name__, model_names)
 
 
+@lru_cache(maxsize = 1)
+def get_static_face_template() -> Tuple[VisionFrame, BoundingBox]:
+	face_template_path = get_face_template_options().get('sources').get('face_template').get('path')
+	face_template = numpy.load(face_template_path)
+	face_template_bounding_box = get_one_face(get_many_faces([ face_template ])).bounding_box
+	return face_template, face_template_bounding_box
+
+
+def collect_model_downloads() -> Tuple[DownloadSet, DownloadSet]:
+	model_set = create_static_model_set('full')
+	model_hash_set = {}
+	model_source_set = {}
+
+	current_model = state_manager.get_item('lip_syncer_model')
+	for lip_syncer_model in [ 'wav2lip_96', 'wav2lip_gan_96', 'edtalk_256' ]:
+		if current_model == lip_syncer_model:
+			model_hash_set['lip_syncer'] = model_set.get(lip_syncer_model).get('hashes').get('lip_syncer')
+			model_source_set['lip_syncer'] = model_set.get(lip_syncer_model).get('sources').get('lip_syncer')
+
+	if state_manager.get_item('lip_syncer_pure_motion') > 0:
+		model_hashes = model_set.get('live_portrait').get('hashes')
+		model_sources = model_set.get('live_portrait').get('sources')
+
+		for model_hash in model_hashes.keys():
+			model_hash_set[model_hash] = model_hashes.get(model_hash)
+
+		for model_source in model_sources.keys():
+			model_source_set[model_source] = model_sources.get(model_source)
+
+	return model_hash_set, model_source_set
+
+
 def get_model_options() -> ModelOptions:
 	model_name = state_manager.get_item('lip_syncer_model')
 	return create_static_model_set('full').get(model_name)
 
 
+def get_face_template_options() -> ModelOptions:
+	template_name = 'face_template'
+	return create_static_model_set('full').get(template_name)
+
+
 def register_args(program : ArgumentParser) -> None:
 	group_processors = find_argument_group(program, 'processors')
 	if group_processors:
 		group_processors.add_argument('--lip-syncer-model', help = wording.get('help.lip_syncer_model'), default = config.get_str_value('processors', 'lip_syncer_model', 'wav2lip_gan_96'), choices = processors_choices.lip_syncer_models)
+		group_processors.add_argument('--lip-syncer-pure-motion', help = wording.get('help.lip_syncer_pure_motion'), type = float, default = config.get_int_value('processors', 'lip_syncer_pure_motion_range', '0'), choices = processors_choices.lip_syncer_pure_motion_range, metavar = create_float_metavar(processors_choices.lip_syncer_pure_motion_range))
 		group_processors.add_argument('--lip-syncer-weight', help = wording.get('help.lip_syncer_weight'), type = float, default = config.get_float_value('processors', 'lip_syncer_weight', '0.5'), choices = processors_choices.lip_syncer_weight_range, metavar = create_float_metavar(processors_choices.lip_syncer_weight_range))
-		facefusion.jobs.job_store.register_step_keys([ 'lip_syncer_model', 'lip_syncer_weight' ])
+		facefusion.jobs.job_store.register_step_keys([ 'lip_syncer_model', 'lip_syncer_pure_motion', 'lip_syncer_weight' ])
 
 
 def apply_args(args : Args, apply_state_item : ApplyStateItem) -> None:
 	apply_state_item('lip_syncer_model', args.get('lip_syncer_model'))
+	apply_state_item('lip_syncer_pure_motion', args.get('lip_syncer_pure_motion'))
 	apply_state_item('lip_syncer_weight', args.get('lip_syncer_weight'))
 
 
 def pre_check() -> bool:
-	model_hash_set = get_model_options().get('hashes')
-	model_source_set = get_model_options().get('sources')
+	model_hash_set, model_source_set = collect_model_downloads()
+	face_template_hash_set = get_face_template_options().get('hashes')
+	face_template_source_set = get_face_template_options().get('sources')
 
-	return conditional_download_hashes(model_hash_set) and conditional_download_sources(model_source_set)
+	return conditional_download_hashes(model_hash_set) and conditional_download_sources(model_source_set) and conditional_download_hashes(face_template_hash_set) and conditional_download_sources(face_template_source_set)
 
 
 def pre_process(mode : ProcessMode) -> bool:
@@ -167,38 +297,198 @@
 
 
 def sync_lip(target_face : Face, temp_audio_frame : AudioFrame, temp_vision_frame : VisionFrame) -> VisionFrame:
-	model_name = state_manager.get_item('lip_syncer_model')
-	model_size = get_model_options().get('size')
 	temp_audio_frame = prepare_audio_frame(temp_audio_frame)
 	crop_vision_frame, affine_matrix = warp_face_by_face_landmark_5(temp_vision_frame, target_face.landmark_set.get('5/68'), 'ffhq_512', (512, 512))
-	crop_masks = []
+	crop_masks = create_occlusion_masks(crop_vision_frame)
+
+	if has_pure_motion():
+		crop_vision_frame, additional_mask = process_live_portrait_motion(target_face, temp_audio_frame, crop_vision_frame, affine_matrix)
+		crop_masks.extend(additional_mask)
+	else:
+		crop_vision_frame, additional_mask = process_standard_lip_sync(target_face, temp_audio_frame, crop_vision_frame, affine_matrix)
+		crop_masks.extend(additional_mask)
 
-	if 'occlusion' in state_manager.get_item('face_mask_types'):
-		occlusion_mask = create_occlusion_mask(crop_vision_frame)
-		crop_masks.append(occlusion_mask)
+	crop_mask = numpy.minimum.reduce(crop_masks)
+	paste_vision_frame = paste_back(temp_vision_frame, crop_vision_frame, crop_mask, affine_matrix)
+	return paste_vision_frame
+
+
+def apply_lip_syncer(crop_vision_frame : VisionFrame, temp_audio_frame : AudioFrame, bounding_box : BoundingBox) -> VisionFrame:
+	model_name = state_manager.get_item('lip_syncer_model')
+	model_size = get_model_options().get('size')
 
 	if model_name == 'edtalk_256':
 		lip_syncer_weight = numpy.array([ state_manager.get_item('lip_syncer_weight') ]).astype(numpy.float32) * 1.25
-		box_mask = create_box_mask(crop_vision_frame, state_manager.get_item('face_mask_blur'), state_manager.get_item('face_mask_padding'))
-		crop_masks.append(box_mask)
 		crop_vision_frame = prepare_crop_frame(crop_vision_frame)
 		crop_vision_frame = forward_edtalk(temp_audio_frame, crop_vision_frame, lip_syncer_weight)
 		crop_vision_frame = normalize_crop_frame(crop_vision_frame)
 	if model_name.startswith('wav2lip'):
-		face_landmark_68 = cv2.transform(target_face.landmark_set.get('68').reshape(1, -1, 2), affine_matrix).reshape(-1, 2)
-		area_mask = create_area_mask(crop_vision_frame, face_landmark_68, [ 'lower-face' ])
-		crop_masks.append(area_mask)
-		bounding_box = create_bounding_box(face_landmark_68)
 		bounding_box = resize_bounding_box(bounding_box, 1 / 8)
 		area_vision_frame, area_matrix = warp_face_by_bounding_box(crop_vision_frame, bounding_box, model_size)
 		area_vision_frame = prepare_crop_frame(area_vision_frame)
 		area_vision_frame = forward_wav2lip(temp_audio_frame, area_vision_frame)
 		area_vision_frame = normalize_crop_frame(area_vision_frame)
 		crop_vision_frame = cv2.warpAffine(area_vision_frame, cv2.invertAffineTransform(area_matrix), (512, 512), borderMode=cv2.BORDER_REPLICATE)
+	return crop_vision_frame
+
+
+def has_pure_motion() -> bool:
+	return state_manager.get_item('lip_syncer_pure_motion') > 0
+
+
+def create_occlusion_masks(crop_vision_frame : VisionFrame) -> List:
+	crop_masks = []
+	if 'occlusion' in state_manager.get_item('face_mask_types'):
+		occlusion_mask = create_occlusion_mask(crop_vision_frame)
+		crop_masks.append(occlusion_mask)
+	return crop_masks
+
+
+def process_live_portrait_motion(
+	target_face : Face,
+	temp_audio_frame : AudioFrame,
+	crop_vision_frame : VisionFrame,
+	affine_matrix
+) -> Tuple[VisionFrame, List]:
+	face_template_expression = extract_template_expression(temp_audio_frame)
+	crop_vision_frame = prepare_refine_frame(crop_vision_frame)
+	feature_volume = forward_extract_feature(crop_vision_frame)
+	pitch, yaw, roll, scale, translation, expression, motion_points = forward_extract_motion(
+		crop_vision_frame
+	)
+
+	motion_points_target = calculate_target_motion_points(
+		pitch, yaw, roll, scale, translation, expression, motion_points
+	)
+	blended_expression = create_blended_expression(expression, face_template_expression)
+	motion_points_source = calculate_source_motion_points(
+		pitch, yaw, roll, scale, translation, blended_expression, motion_points
+	)
+	motion_points_source = forward_stitch_motion_points(
+		motion_points_source, motion_points_target
+	)
+
+	crop_vision_frame = forward_generate_frame(
+		feature_volume, motion_points_source, motion_points_target
+	)
+	crop_vision_frame = normalize_refine_frame(crop_vision_frame)
+	box_mask = create_box_mask(
+		crop_vision_frame,
+		state_manager.get_item('face_mask_blur'),
+		state_manager.get_item('face_mask_padding')
+	)
+	return crop_vision_frame, [box_mask]
+
+
+def process_standard_lip_sync(
+	target_face : Face,
+	temp_audio_frame : AudioFrame,
+	crop_vision_frame : VisionFrame,
+	affine_matrix
+) -> Tuple[VisionFrame, List]:
+	model_name = state_manager.get_item('lip_syncer_model')
+	face_landmark_68 = cv2.transform(
+		target_face.landmark_set.get('68').reshape(1, -1, 2), affine_matrix
+	).reshape(-1, 2)
+	bounding_box = create_bounding_box(face_landmark_68)
+	crop_vision_frame = apply_lip_syncer(crop_vision_frame, temp_audio_frame, bounding_box)
+
+	masks = []
+	if model_name == 'edtalk_256':
+		box_mask = create_box_mask(
+			crop_vision_frame,
+			state_manager.get_item('face_mask_blur'),
+			state_manager.get_item('face_mask_padding')
+		)
+		masks.append(box_mask)
+	if model_name.startswith('wav2lip'):
+		area_mask = create_area_mask(
+			crop_vision_frame, face_landmark_68, [ 'lower-face' ]
+		)
+		masks.append(area_mask)
+	return crop_vision_frame, masks
+
+
+def extract_template_expression(temp_audio_frame : AudioFrame) -> LivePortraitExpression:
+	face_template_frame, face_template_bounding_box = get_static_face_template()
+	face_template_crop_frame = apply_lip_syncer(
+		face_template_frame.copy(), temp_audio_frame, face_template_bounding_box.copy()
+	)
+	face_template_crop_frame = prepare_refine_frame(face_template_crop_frame)
+	face_template_expression = forward_extract_motion(face_template_crop_frame)[5]
+	face_template_expression *= state_manager.get_item('lip_syncer_pure_motion')
+	return face_template_expression
+
+
+def calculate_target_motion_points(
+	pitch, yaw, roll, scale, translation, expression, motion_points
+) -> LivePortraitMotionPoints:
+	rotation = create_rotation(pitch, yaw, roll)
+	return scale * (motion_points @ rotation.T + expression) + translation
+
 
-	crop_mask = numpy.minimum.reduce(crop_masks)
-	paste_vision_frame = paste_back(temp_vision_frame, crop_vision_frame, crop_mask, affine_matrix)
-	return paste_vision_frame
+def calculate_source_motion_points(
+	pitch, yaw, roll, scale, translation, expression, motion_points
+) -> LivePortraitMotionPoints:
+	rotation = create_rotation(pitch, yaw, roll)
+	motion_points_source = motion_points @ rotation.T
+	motion_points_source += expression
+	motion_points_source *= scale
+	motion_points_source += translation
+	return motion_points_source
+
+
+def create_blended_expression(
+	expression : LivePortraitExpression,
+	face_template_expression : LivePortraitExpression
+) -> LivePortraitExpression:
+	temp_expression = expression.copy()
+	temp_expression = blend_expression(
+		6, temp_expression, face_template_expression, 0.5, 0.5, 0.5
+	)
+	temp_expression = blend_expression(
+		12, temp_expression, face_template_expression, 0.5, 0.5, 0.5
+	)
+	temp_expression = blend_expression(
+		14, temp_expression, face_template_expression, 0.6, 0.7, 0.7
+	)
+	temp_expression = blend_expression(
+		17, temp_expression, face_template_expression, 0.5, 0.8, 0.7
+	)
+	temp_expression = blend_expression(
+		20, temp_expression, face_template_expression, 0.5, 0.6, 0.7
+	)
+	lip_open = float(numpy.interp(
+		face_template_expression[0, 19, 1].clip(0, 1), [0, 1], [0.9, 1.5]
+	))
+	temp_expression = blend_expression(
+		19, temp_expression, face_template_expression, 0.5, lip_open, 0.85
+	)
+	temp_expression = limit_expression(temp_expression)
+	return temp_expression
+
+
+def blend_expression(
+	index : int,
+	temp_expression : LivePortraitExpression,
+	face_template_expression : LivePortraitExpression,
+	blend_x : float,
+	blend_y : float,
+	blend_z : float
+) -> LivePortraitExpression:
+	temp_expression[0, index, 0] = (
+		temp_expression[0, index, 0] * max(1 - blend_x, 0) +
+		face_template_expression[0, index, 0] * blend_x
+	)
+	temp_expression[0, index, 1] = (
+		temp_expression[0, index, 1] * max(1 - blend_y, 0) +
+		face_template_expression[0, index, 1] * blend_y
+	)
+	temp_expression[0, index, 2] = (
+		temp_expression[0, index, 2] * max(1 - blend_z, 0) +
+		face_template_expression[0, index, 2] * blend_z
+	)
+	return temp_expression
 
 
 def forward_wav2lip(temp_audio_frame : AudioFrame, area_vision_frame : VisionFrame) -> VisionFrame:
@@ -228,6 +518,57 @@
 	return crop_vision_frame
 
 
+def forward_extract_feature(crop_vision_frame : VisionFrame) -> LivePortraitFeatureVolume:
+	feature_extractor = get_inference_pool().get('feature_extractor')
+
+	with conditional_thread_semaphore():
+		feature_volume = feature_extractor.run(None,
+		{
+			'input': crop_vision_frame
+		})[0]
+
+	return feature_volume
+
+
+def forward_extract_motion(crop_vision_frame : VisionFrame) -> Tuple[LivePortraitPitch, LivePortraitYaw, LivePortraitRoll, LivePortraitScale, LivePortraitTranslation, LivePortraitExpression, LivePortraitMotionPoints]:
+	motion_extractor = get_inference_pool().get('motion_extractor')
+
+	with conditional_thread_semaphore():
+		pitch, yaw, roll, scale, translation, expression, motion_points = motion_extractor.run(None,
+		{
+			'input': crop_vision_frame
+		})
+
+	return pitch, yaw, roll, scale, translation, expression, motion_points
+
+
+def forward_stitch_motion_points(source_motion_points : LivePortraitMotionPoints, target_motion_points : LivePortraitMotionPoints) -> LivePortraitMotionPoints:
+	stitcher = get_inference_pool().get('stitcher')
+
+	with thread_semaphore():
+		motion_points = stitcher.run(None,
+		{
+			'source': source_motion_points,
+			'target': target_motion_points
+		})[0]
+
+	return motion_points
+
+
+def forward_generate_frame(feature_volume : LivePortraitFeatureVolume, source_motion_points : LivePortraitMotionPoints, target_motion_points : LivePortraitMotionPoints) -> VisionFrame:
+	generator = get_inference_pool().get('generator')
+
+	with thread_semaphore():
+		crop_vision_frame = generator.run(None,
+		{
+			'feature_volume': feature_volume,
+			'source': source_motion_points,
+			'target': target_motion_points
+		})[0][0]
+
+	return crop_vision_frame
+
+
 def prepare_audio_frame(temp_audio_frame : AudioFrame) -> AudioFrame:
 	temp_audio_frame = numpy.maximum(numpy.exp(-5 * numpy.log(10)), temp_audio_frame)
 	temp_audio_frame = numpy.log10(temp_audio_frame) * 1.6 + 3.2
@@ -242,7 +583,7 @@
 	model_size = get_model_options().get('size')
 
 	if model_type == 'edtalk':
-		crop_vision_frame = cv2.resize(crop_vision_frame, (256, 256), interpolation = cv2.INTER_AREA)
+		crop_vision_frame = cv2.resize(crop_vision_frame, model_size, interpolation = cv2.INTER_AREA)
 		crop_vision_frame = crop_vision_frame[:, :, ::-1] / 255.0
 		crop_vision_frame = numpy.expand_dims(crop_vision_frame.transpose(2, 0, 1), axis = 0).astype(numpy.float32)
 	if model_type == 'wav2lip':
@@ -255,6 +596,13 @@
 	return crop_vision_frame
 
 
+def prepare_refine_frame(crop_vision_frame : VisionFrame) -> VisionFrame:
+	crop_vision_frame = cv2.resize(crop_vision_frame, (256, 256), interpolation = cv2.INTER_CUBIC)
+	crop_vision_frame = crop_vision_frame[:, :, ::-1] / 255.0
+	crop_vision_frame = numpy.expand_dims(crop_vision_frame.transpose(2, 0, 1), axis=0).astype(numpy.float32)
+	return crop_vision_frame
+
+
 def resize_bounding_box(bounding_box : BoundingBox, aspect_ratio : float) -> BoundingBox:
 	x1, y1, x2, y2 = bounding_box
 	y1 -= numpy.abs(y2 - y1) * aspect_ratio
@@ -272,6 +620,15 @@
 		crop_vision_frame = crop_vision_frame[:, :, ::-1]
 		crop_vision_frame = cv2.resize(crop_vision_frame, (512, 512), interpolation=cv2.INTER_CUBIC)
 
+	return crop_vision_frame
+
+
+def normalize_refine_frame(crop_vision_frame : VisionFrame) -> VisionFrame:
+	crop_vision_frame = crop_vision_frame.transpose(1, 2, 0)
+	crop_vision_frame = crop_vision_frame.clip(0, 1) * 255
+	crop_vision_frame = crop_vision_frame.astype(numpy.uint8)
+	crop_vision_frame = crop_vision_frame[:, :, ::-1]
+	crop_vision_frame = cv2.resize(crop_vision_frame, (512, 512), interpolation = cv2.INTER_CUBIC)
 	return crop_vision_frame
 
 
Index: facefusion/processors/types.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/facefusion/processors/types.py b/facefusion/processors/types.py
--- a/facefusion/processors/types.py	(revision d1385fed0d45329171c2119de5d3713a09d18399)
+++ b/facefusion/processors/types.py	(date 1750151355189)
@@ -104,7 +104,8 @@
 	'frame_enhancer_model',
 	'frame_enhancer_blend',
 	'lip_syncer_model',
-	'lip_syncer_weight'
+	'lip_syncer_weight',
+	'lip_syncer_pure_motion'
 ]
 ProcessorState = TypedDict('ProcessorState',
 {
@@ -140,7 +141,8 @@
 	'frame_colorizer_blend' : int,
 	'frame_enhancer_model' : FrameEnhancerModel,
 	'frame_enhancer_blend' : int,
-	'lip_syncer_model' : LipSyncerModel
+	'lip_syncer_model' : LipSyncerModel,
+	'lip_syncer_pure_motion': float
 })
 ProcessorStateSet : TypeAlias = Dict[AppContext, ProcessorState]
 
Index: facefusion/uis/types.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/facefusion/uis/types.py b/facefusion/uis/types.py
--- a/facefusion/uis/types.py	(revision d1385fed0d45329171c2119de5d3713a09d18399)
+++ b/facefusion/uis/types.py	(date 1750151355199)
@@ -60,6 +60,7 @@
 	'frame_enhancer_model_dropdown',
 	'job_list_job_status_checkbox_group',
 	'lip_syncer_model_dropdown',
+	'lip_syncer_pure_motion_slider',
 	'lip_syncer_weight_slider',
 	'output_image',
 	'output_video',
Index: facefusion/processors/choices.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/facefusion/processors/choices.py b/facefusion/processors/choices.py
--- a/facefusion/processors/choices.py	(revision d1385fed0d45329171c2119de5d3713a09d18399)
+++ b/facefusion/processors/choices.py	(date 1750151355186)
@@ -219,3 +219,4 @@
 frame_colorizer_blend_range : Sequence[int] = create_int_range(0, 100, 1)
 frame_enhancer_blend_range : Sequence[int] = create_int_range(0, 100, 1)
 lip_syncer_weight_range : Sequence[float] = create_float_range(0.0, 1.0, 0.05)
+lip_syncer_pure_motion_range : Sequence[float] = create_float_range(0.0, 1.5, 0.25)
